p2panda-net/src/network.rs

```diff
@@ -555,19 +576,239 @@ async fn handle_connection(
     }
 }
 
+#[cfg(test)]
+mod sync_protocols {
+    use std::sync::Arc;
+
+    use async_trait::async_trait;
+    use futures_lite::{AsyncRead, AsyncWrite, StreamExt};
+    use futures_util::{Sink, SinkExt};
+    use p2panda_sync::protocols::utils::{into_sink, into_stream};
+    use p2panda_sync::traits::SyncProtocol;
+    use p2panda_sync::{AppMessage, SyncError};
+    use serde::{Deserialize, Serialize};
+    use tracing::debug;
+
+    use crate::TopicId;
+
+    #[derive(Debug, Serialize, Deserialize)]
+    enum DummyProtocolMessage {
+        Topic(TopicId),
+        Done,
+    }
+
+    /// A sync implementation which fulfills basic protocol requirements but nothing more  
+    #[derive(Debug)]
+    pub struct DummyProtocol {}
+
+    #[async_trait]
+    impl<'a> SyncProtocol<'a> for DummyProtocol {
+        fn name(&self) -> &'static str {
+            static DUMMY_PROTOCOL_NAME: &str = "dummy_protocol";
+            DUMMY_PROTOCOL_NAME
+        }
+        async fn open(
+            self: Arc<Self>,
+            topic: &TopicId,
+            tx: Box<&'a mut (dyn AsyncWrite + Send + Unpin)>,
+            rx: Box<&'a mut (dyn AsyncRead + Send + Unpin)>,
+            mut app_tx: Box<&'a mut (dyn Sink<AppMessage, Error = SyncError> + Send + Unpin)>,
+        ) -> Result<(), SyncError> {
+            debug!("DummyProtocol: open sync session");
+
+            let mut sink = into_sink(tx);
+            let mut stream = into_stream(rx);
+
+            sink.send(DummyProtocolMessage::Topic(*topic)).await?;
+            sink.send(DummyProtocolMessage::Done).await?;
+            app_tx.send(AppMessage::Topic(*topic)).await?;
+
+            while let Some(result) = stream.next().await {
+                let message: DummyProtocolMessage = result?;
+                debug!("message received: {:?}", message);
+
+                match &message {
+                    DummyProtocolMessage::Topic(_) => panic!(),
+                    DummyProtocolMessage::Done => break,
+                }
+            }
+
+            sink.flush().await?;
+            sink.close().await?;
+
+            app_tx.flush().await?;
+            app_tx.close().await?;
```

Comment:
Wondering if this little dance shouldn't be handled always for every sync implementation inside the caller / handler one level up

Reply:
I experimented with that but some new errors appeared, I'll raise an issue and deal with it once we have the connection handler in place if that's ok :pray:

Reply:
https://github.com/p2panda/p2panda/issues/557